### ISSUE TITLE
[BUG]: How to use tools in dynamo vllm v0.2.0

### ISSUE THREAD

**Issue by batuhanunverdi** [ai-dynamo/dynamo#1808]
### Describe the Bug

I'm using dynamo v0.2.0 and I'm running the Qwen/Qwen3-32B model in aggregated mode with tensor-parallel-size 2 - 1 worker 2 gpu. and I also use  

> --enable-auto-tool-choice --tool-call-parser hermes

 arguments. but after that I got an error **unrecognized arguments:--enable-auto-tool-choice --tool-call-parser hermes** and it shows the arguments that I can use in dynamo serve. 
My Question is how can I use the tools in dynamo 

> usage: serve_dynamo.py [-h] [--router {random,round-robin,kv,kv-load}] [--remote-prefill] [--conditional-disagg] [--max-local-prefill-length MAX_LOCAL_PREFILL_LENGTH]
                       [--max-prefill-queue-size MAX_PREFILL_QUEUE_SIZE] [--model MODEL] [--task {auto,generate,embedding,embed,classify,score,reward,transcription}] [--tokenizer TOKENIZER]
                       [--hf-config-path HF_CONFIG_PATH] [--skip-tokenizer-init] [--revision REVISION] [--code-revision CODE_REVISION] [--tokenizer-revision TOKENIZER_REVISION]
                       [--tokenizer-mode {auto,slow,mistral,custom}] [--trust-remote-code] [--allowed-local-media-path ALLOWED_LOCAL_MEDIA_PATH]
                       [--load-format {auto,pt,safetensors,npcache,dummy,tensorizer,sharded_state,gguf,bitsandbytes,mistral,runai_streamer,fastsafetensors}] [--download-dir DOWNLOAD_DIR]
                       [--model-loader-extra-config MODEL_LOADER_EXTRA_CONFIG] [--use-tqdm-on-load | --no-use-tqdm-on-load] [--config-format {auto,hf,mistral}]
                       [--dtype {auto,half,float16,bfloat16,float,float32}] [--kv-cache-dtype {auto,fp8,fp8_e5m2,fp8_e4m3}] [--max-model-len MAX_MODEL_LEN]
                       [--guided-decoding-backend GUIDED_DECODING_BACKEND] [--logits-processor-pattern LOGITS_PROCESSOR_PATTERN] [--model-impl {auto,vllm,transformers}]
                       [--distributed-executor-backend {ray,mp,uni,external_launcher}] [--pipeline-parallel-size PIPELINE_PARALLEL_SIZE] [--tensor-parallel-size TENSOR_PARALLEL_SIZE]
                       [--data-parallel-size DATA_PARALLEL_SIZE] [--enable-expert-parallel] [--max-parallel-loading-workers MAX_PARALLEL_LOADING_WORKERS] [--ray-workers-use-nsight]
                       [--disable-custom-all-reduce] [--block-size {8,16,32,64,128}] [--enable-prefix-caching | --no-enable-prefix-caching] [--prefix-caching-hash-algo {builtin,sha256}]
                       [--disable-sliding-window] [--use-v2-block-manager] [--num-lookahead-slots NUM_LOOKAHEAD_SLOTS] [--seed SEED] [--swap-space SWAP_SPACE] [--cpu-offload-gb CPU_OFFLOAD_GB]
                       [--gpu-memory-utilization GPU_MEMORY_UTILIZATION] [--num-gpu-blocks-override NUM_GPU_BLOCKS_OVERRIDE] [--max-num-batched-tokens MAX_NUM_BATCHED_TOKENS]
                       [--max-num-partial-prefills MAX_NUM_PARTIAL_PREFILLS] [--max-long-partial-prefills MAX_LONG_PARTIAL_PREFILLS] [--long-prefill-token-threshold LONG_PREFILL_TOKEN_THRESHOLD]
                       [--max-num-seqs MAX_NUM_SEQS] [--max-logprobs MAX_LOGPROBS] [--disable-log-stats]
                       [--quantization {aqlm,awq,deepspeedfp,tpu_int8,fp8,ptpc_fp8,fbgemm_fp8,modelopt,nvfp4,marlin,gguf,gptq_marlin_24,gptq_marlin,awq_marlin,gptq,compressed-tensors,bitsandbytes,qqq,hqq,experts_int8,neuron_quant,ipex,quark,moe_wna16,torchao,None}]
                       [--rope-scaling ROPE_SCALING] [--rope-theta ROPE_THETA] [--hf-token [HF_TOKEN]] [--hf-overrides HF_OVERRIDES] [--enforce-eager] [--max-seq-len-to-capture MAX_SEQ_LEN_TO_CAPTURE]
                       [--tokenizer-pool-size TOKENIZER_POOL_SIZE] [--tokenizer-pool-type TOKENIZER_POOL_TYPE] [--tokenizer-pool-extra-config TOKENIZER_POOL_EXTRA_CONFIG]
                       [--limit-mm-per-prompt LIMIT_MM_PER_PROMPT] [--mm-processor-kwargs MM_PROCESSOR_KWARGS] [--disable-mm-preprocessor-cache] [--enable-lora] [--enable-lora-bias]
                       [--max-loras MAX_LORAS] [--max-lora-rank MAX_LORA_RANK] [--lora-extra-vocab-size LORA_EXTRA_VOCAB_SIZE] [--lora-dtype {auto,float16,bfloat16}]
                       [--long-lora-scaling-factors LONG_LORA_SCALING_FACTORS] [--max-cpu-loras MAX_CPU_LORAS] [--fully-sharded-loras] [--enable-prompt-adapter]
                       [--max-prompt-adapters MAX_PROMPT_ADAPTERS] [--max-prompt-adapter-token MAX_PROMPT_ADAPTER_TOKEN] [--device {auto,cuda,neuron,cpu,tpu,xpu,hpu}]
                       [--num-scheduler-steps NUM_SCHEDULER_STEPS] [--multi-step-stream-outputs [MULTI_STEP_STREAM_OUTPUTS]] [--scheduler-delay-factor SCHEDULER_DELAY_FACTOR]
                       [--enable-chunked-prefill [ENABLE_CHUNKED_PREFILL]] [--speculative-config SPECULATIVE_CONFIG] [--ignore-patterns IGNORE_PATTERNS] [--preemption-mode PREEMPTION_MODE]
                       [--served-model-name SERVED_MODEL_NAME [SERVED_MODEL_NAME ...]] [--qlora-adapter-name-or-path QLORA_ADAPTER_NAME_OR_PATH]
                       [--show-hidden-metrics-for-version SHOW_HIDDEN_METRICS_FOR_VERSION] [--otlp-traces-endpoint OTLP_TRACES_ENDPOINT] [--collect-detailed-traces COLLECT_DETAILED_TRACES]
                       [--disable-async-output-proc] [--scheduling-policy {fcfs,priority}] [--scheduler-cls SCHEDULER_CLS] [--override-neuron-config OVERRIDE_NEURON_CONFIG]
                       [--override-pooler-config OVERRIDE_POOLER_CONFIG] [--compilation-config COMPILATION_CONFIG] [--kv-transfer-config KV_TRANSFER_CONFIG] [--worker-cls WORKER_CLS]
                       [--worker-extension-cls WORKER_EXTENSION_CLS] [--generation-config GENERATION_CONFIG] [--override-generation-config OVERRIDE_GENERATION_CONFIG] [--enable-sleep-mode]
                       [--calculate-kv-scales] [--additional-config ADDITIONAL_CONFIG] [--enable-reasoning] [--reasoning-parser {deepseek_r1,granite}] [--disable-cascade-attn]
                       [--disable-chunked-mm-input [DISABLE_CHUNKED_MM_INPUT]] [--disable-log-requests]

### Steps to Reproduce

I follow the example/llm steps and only change the yaml config.

### Expected Behavior

I thought I can use tools in dynamo

### Actual Behavior

I can not use it.

### Environment

RTX ADA6000 x2

**Comment by ishandhanani** [comment]
Hi @batuhanunverdi - thank you for your interested in tool use! We just merged in support in our frontend API for unary tool use (https://github.com/ai-dynamo/dynamo/pull/1800). However - I am not sure why it's not being passed through to the engine itself. I see that you're using the `llm/` examples which are older and in the process of being deprecated. Can you try with vllm_v1? 

Additionally there is a WIP PR (https://github.com/ai-dynamo/dynamo/pull/1756) that should make it much easier to pass in vllm flags when running

**Comment by batuhanunverdi** [comment]
Hİ @ishandhanani - thanks for your reply. I'll try it and get back to you immediately.

**Comment by xidiancpy** [comment]
> Hİ [@ishandhanani](https://github.com/ishandhanani) - thanks for your reply. I'll try it and get back to you immediately.

Could you share one of your container with me?

**Comment by batuhanunverdi** [comment]
Hi  @ishandhanani 
I built an image using the --framework `VLLM_V1` flag with the `v0.3.2 `tag. I used the run.sh script and executed the following command:

> dynamo serve graphs.agg:Frontend -f configs/agg.yaml \
  --Common.model=Qwen/Qwen2.5-0.5B \
  --Common.served_model_name=Qwen/Qwen2.5-0.5B \
  --Frontend.served_model_name=Qwen/Qwen2.5-0.5B \
  --VllmDecodeWorker.enforce-eager=true \
  --VllmDecodeWorker.enable-prefix-caching=true \
  --VllmDecodeWorker.block-size=64 \
  --VllmDecodeWorker.max-model-len=2048 \
  --VllmDecodeWorker.max-num-batched-tokens=2048 \
  --VllmDecodeWorker.ServiceArgs.workers=1 \
  --VllmDecodeWorker.ServiceArgs.resources.gpu=1 \
  --VllmDecodeWorker.enable-auto-tool-choice=true \
  --VllmDecodeWorker.tool-call-parser=hermes


However, I'm still getting the same error: unrecognized arguments.
Could you please advise what might be causing this or if there's something wrong with the arguments?

Thanks in advance!

**Comment by xidiancpy** [comment]
> Hi [@ishandhanani](https://github.com/ishandhanani) I built an image using the --framework `VLLM_V1` flag with the `v0.3.2 `tag. I used the run.sh script and executed the following command:
> 
> > dynamo serve graphs.agg:Frontend -f configs/agg.yaml 
> >   --Common.model=Qwen/Qwen2.5-0.5B 
> >   --Common.served_model_name=Qwen/Qwen2.5-0.5B 
> >   --Frontend.served_model_name=Qwen/Qwen2.5-0.5B 
> >   --VllmDecodeWorker.enforce-eager=true 
> >   --VllmDecodeWorker.enable-prefix-caching=true 
> >   --VllmDecodeWorker.block-size=64 
> >   --VllmDecodeWorker.max-model-len=2048 
> >   --VllmDecodeWorker.max-num-batched-tokens=2048 
> >   --VllmDecodeWorker.ServiceArgs.workers=1 
> >   --VllmDecodeWorker.ServiceArgs.resources.gpu=1 
> >   --VllmDecodeWorker.enable-auto-tool-choice=true 
> >   --VllmDecodeWorker.tool-call-parser=hermes
> 
> However, I'm still getting the same error: unrecognized arguments. Could you please advise what might be causing this or if there's something wrong with the arguments?
> 
> Thanks in advance!

Could you share  your image with me?

**Comment by batuhanunverdi** [comment]
@xidiancpy Hİ. I pushed my image to docker hub.
https://hub.docker.com/repository/docker/mbatuhan/dynamo-base/tags/0.3.2_v1/sha256:fa4b81ee03a1bba135f56bf13fc29ea25ed32384e36254e1e28e41501409e2c8

I created tihs image with these commands. 

> ./container/build.sh --framework VLLM_V1 --target dev
> docker tag dynamo:latest-vllm <your-registry>/dynamo-base:latest-vllm
> docker login <your-registry>
> docker push <your-registry>/dynamo-base:latest-vllm

**Comment by ishandhanani** [comment]
We've now added unary tool use in our API. Our `0.4.0` release should have an initial implementation of this! It should be out in the next week or so. Please feel free to reopen if this is not solved with it!

**Comment by libin817927** [comment]
> We've now added unary tool use in our API. Our `0.4.0` release should have an initial implementation of this! It should be out in the next week or so. Please feel free to reopen if this is not solved with it!

@ishandhanani Could you tell me how to use it? Is there any documentation available?

**Comment by matheus-1618** [comment]
Any docs on how to use this with SGlang?
